File: library/src/com/google/maps/android/MarkerManager.java
/*
 * Copyright 2013 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.maps.android;

import android.view.View;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Keeps track of collections of markers on the map. Delegates all Marker-related events to each
 * collection's individually managed listeners.
 * <p/>
 * All marker operations (adds and removes) should occur via its collection class. That is, don't
 * add a marker via a collection, then remove it via Marker.remove()
 */
public class MarkerManager implements
        GoogleMap.OnInfoWindowClickListener,
        GoogleMap.OnMarkerClickListener,
        GoogleMap.OnMarkerDragListener,
        GoogleMap.InfoWindowAdapter {

    private final GoogleMap mMap;

    private final Map<String, Collection> mNamedCollections = new HashMap<String, Collection>();
    private final Map<Marker, Collection> mAllMarkers = new HashMap<Marker, Collection>();

    public MarkerManager(GoogleMap map) {
        mMap = map;
    }

    public Collection newCollection() {
        return new Collection();
    }

    /**
     * Create a new named collection, which can later be looked up by {@link #getCollection(String)}
     *
     * @param id a unique id for this collection.
     */
    public Collection newCollection(String id) {
        if (mNamedCollections.get(id) != null) {
            throw new IllegalArgumentException("collection id is not unique: " + id);
        }
        Collection collection = new Collection();
        mNamedCollections.put(id, collection);
        return collection;
    }

    /**
     * Gets a named collection that was created by {@link #newCollection(String)}
     *
     * @param id the unique id for this collection.
     */
    public Collection getCollection(String id) {
        return mNamedCollections.get(id);
    }

    @Override
    public View getInfoWindow(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        if (collection != null && collection.mInfoWindowAdapter != null) {
            return collection.mInfoWindowAdapter.getInfoWindow(marker);
        }
        return null;
    }

    @Override
    public View getInfoContents(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        if (collection != null && collection.mInfoWindowAdapter != null) {
            return collection.mInfoWindowAdapter.getInfoContents(marker);
        }
        return null;
    }

    @Override
    public void onInfoWindowClick(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        if (collection != null && collection.mInfoWindowClickListener != null) {
            collection.mInfoWindowClickListener.onInfoWindowClick(marker);
        }
    }

    @Override
    public boolean onMarkerClick(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        if (collection != null && collection.mMarkerClickListener != null) {
            return collection.mMarkerClickListener.onMarkerClick(marker);
        }
        return false;
    }

    @Override
    public void onMarkerDragStart(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        if (collection != null && collection.mMarkerDragListener != null) {
            collection.mMarkerDragListener.onMarkerDragStart(marker);
        }
    }

    @Override
    public void onMarkerDrag(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        if (collection != null && collection.mMarkerDragListener != null) {
            collection.mMarkerDragListener.onMarkerDrag(marker);
        }
    }

    @Override
    public void onMarkerDragEnd(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        if (collection != null && collection.mMarkerDragListener != null) {
            collection.mMarkerDragListener.onMarkerDragEnd(marker);
        }
    }

    /**
     * Removes a marker from its collection.
     *
     * @param marker the marker to remove.
     * @return true if the marker was removed.
     */
    public boolean remove(Marker marker) {
        Collection collection = mAllMarkers.get(marker);
        return collection != null && collection.remove(marker);
    }

    public class Collection {
        private final Set<Marker> mMarkers = new HashSet<Marker>();
        private GoogleMap.OnInfoWindowClickListener mInfoWindowClickListener;
        private GoogleMap.OnMarkerClickListener mMarkerClickListener;
        private GoogleMap.OnMarkerDragListener mMarkerDragListener;
        private GoogleMap.InfoWindowAdapter mInfoWindowAdapter;

        public Collection() {
        }

        public Marker addMarker(MarkerOptions opts) {
            Marker marker = mMap.addMarker(opts);
            mMarkers.add(marker);
            mAllMarkers.put(marker, Collection.this);
            return marker;
        }

        public void addAll(java.util.Collection<MarkerOptions> opts) {
            for (MarkerOptions opt : opts) {
                addMarker(opt);
            }
        }

        public void addAll(java.util.Collection<MarkerOptions> opts, boolean defaultVisible) {
            for (MarkerOptions opt : opts) {
                addMarker(opt).setVisible(defaultVisible);
            }
        }

        public void showAll() {
            for (Marker marker : mMarkers) {
                marker.setVisible(true);
            }
        }

        public void hideAll() {
            for (Marker marker : mMarkers) {
                marker.setVisible(false);
            }
        }

        public boolean remove(Marker marker) {
            if (mMarkers.remove(marker)) {
                mAllMarkers.remove(marker);
                marker.remove();
                return true;
            }
            return false;
        }

        public void clear() {
            for (Marker marker : mMarkers) {
                marker.remove();
                mAllMarkers.remove(marker);
            }
            mMarkers.clear();
        }

        public java.util.Collection<Marker> getMarkers() {
            return Collections.unmodifiableCollection(mMarkers);
        }

        public void setOnInfoWindowClickListener(GoogleMap.OnInfoWindowClickListener infoWindowClickListener) {
            mInfoWindowClickListener = infoWindowClickListener;
        }

        public void setOnMarkerClickListener(GoogleMap.OnMarkerClickListener markerClickListener) {
            mMarkerClickListener = markerClickListener;
        }

        public void setOnMarkerDragListener(GoogleMap.OnMarkerDragListener markerDragListener) {
            mMarkerDragListener = markerDragListener;
        }

        public void setOnInfoWindowAdapter(GoogleMap.InfoWindowAdapter infoWindowAdapter) {
            mInfoWindowAdapter = infoWindowAdapter;
        }
    }
}

File: library/src/androidMain/kotlin/com/google/maps/android/collections/MarkerManager.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.maps.android.collections

import android.view.View
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.AdvancedMarkerOptions
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import kotlin.collections.Collection as KotlinCollection

/**
 * Keeps track of collections of markers on the map. Delegates all Marker-related events to each
 * collection's individually managed listeners.
 *
 * All marker operations (adds and removes) should occur via its collection class. That is, don't
 * add a marker via a collection, then remove it via Marker.remove()
 */
open class MarkerManager(map: GoogleMap) :
    MapObjectManager<Marker, MarkerManager.Collection>(map),
    GoogleMap.OnInfoWindowClickListener,
    GoogleMap.OnMarkerClickListener,
    GoogleMap.OnMarkerDragListener,
    GoogleMap.InfoWindowAdapter,
    GoogleMap.OnInfoWindowLongClickListener {

    override fun setListenersOnUiThread() {
        mMap.setOnInfoWindowClickListener(this)
        mMap.setOnInfoWindowLongClickListener(this)
        mMap.setOnMarkerClickListener(this)
        mMap.setOnMarkerDragListener(this)
        mMap.setInfoWindowAdapter(this)
    }

    override fun newCollection(): Collection = Collection()

    override fun getInfoWindow(marker: Marker): View? =
        mAllObjects[marker]?.mInfoWindowAdapter?.getInfoWindow(marker)

    override fun getInfoContents(marker: Marker): View? =
        mAllObjects[marker]?.mInfoWindowAdapter?.getInfoContents(marker)

    override fun onInfoWindowClick(marker: Marker) {
        mAllObjects[marker]?.mInfoWindowClickListener?.onInfoWindowClick(marker)
    }

    override fun onInfoWindowLongClick(marker: Marker) {
        mAllObjects[marker]?.mInfoWindowLongClickListener?.onInfoWindowLongClick(marker)
    }

    override fun onMarkerClick(marker: Marker): Boolean =
        mAllObjects[marker]?.mMarkerClickListener?.onMarkerClick(marker) ?: false

    override fun onMarkerDragStart(marker: Marker) {
        mAllObjects[marker]?.mMarkerDragListener?.onMarkerDragStart(marker)
    }

    override fun onMarkerDrag(marker: Marker) {
        mAllObjects[marker]?.mMarkerDragListener?.onMarkerDrag(marker)
    }

    override fun onMarkerDragEnd(marker: Marker) {
        mAllObjects[marker]?.mMarkerDragListener?.onMarkerDragEnd(marker)
    }

    override fun removeObjectFromMap(marker: Marker) {
        marker.remove()
    }

    override fun setVisible(mapObject: Marker, visible: Boolean) {
        mapObject.isVisible = visible
    }

    /** A collection of [Marker]s on the map with its own set of listeners. */
    open inner class Collection : MapObjectManager<Marker, Collection>.Collection() {
        internal var mInfoWindowClickListener: GoogleMap.OnInfoWindowClickListener? = null
        internal var mInfoWindowLongClickListener: GoogleMap.OnInfoWindowLongClickListener? = null
        internal var mMarkerClickListener: GoogleMap.OnMarkerClickListener? = null
        internal var mMarkerDragListener: GoogleMap.OnMarkerDragListener? = null
        internal var mInfoWindowAdapter: GoogleMap.InfoWindowAdapter? = null

        open fun addMarker(opts: MarkerOptions): Marker =
            checkAndAdd(mMap.addMarker(opts), "Marker")

        open fun addMarker(opts: AdvancedMarkerOptions): Marker =
            checkAndAdd(mMap.addMarker(opts), "AdvancedMarker")

        open fun addAll(opts: KotlinCollection<MarkerOptions>) =
            addAll(opts, ::addMarker)

        open fun addAll(opts: KotlinCollection<MarkerOptions>, defaultVisible: Boolean) =
            addAll(opts, defaultVisible, ::addMarker)

        open fun getMarkers(): KotlinCollection<Marker> = getObjects()

        open fun setOnInfoWindowClickListener(infoWindowClickListener: GoogleMap.OnInfoWindowClickListener?) {
            mInfoWindowClickListener = infoWindowClickListener
        }

        open fun setOnInfoWindowLongClickListener(infoWindowLongClickListener: GoogleMap.OnInfoWindowLongClickListener?) {
            mInfoWindowLongClickListener = infoWindowLongClickListener
        }

        open fun setOnMarkerClickListener(markerClickListener: GoogleMap.OnMarkerClickListener?) {
            mMarkerClickListener = markerClickListener
        }

        open fun setOnMarkerDragListener(markerDragListener: GoogleMap.OnMarkerDragListener?) {
            mMarkerDragListener = markerDragListener
        }

        open fun setInfoWindowAdapter(infoWindowAdapter: GoogleMap.InfoWindowAdapter?) {
            mInfoWindowAdapter = infoWindowAdapter
        }
    }
}
